### PR TITLE
Expand support of Azure functions to include jdk 17 

### LINKED ISSUE #1629: Azure functions support 17 but we limit 8 and 11
> The Azure team has publically released the Java 17 runtime on December 17. 
> Link to the article: https://azure.microsoft.com/en-in/updates/generally-available-azure-functions-support-for-java-17/
> 
> But the micronaut project initializer still shows only Java 8 or 11 are supported.

see: 
https://micronaut.io/launch?type=FUNCTION&name=demo&package=com.example&javaVersion=JDK_17&lang=JAVA&build=GRADLE&test=JUNIT&features=azure-function&version=3.8.5

Java 17 should be a valid option.

### PR DESCRIPTION
**Pull request by wetted**
(currently limited to 8 and 11)

fixes #1629

### CODE REVIEW

**Comment by wetted** [comment]
I have an interesting Spock test problem. I added JDK_17 as a supported JDK for Azure Function feature. That means it supports all the JDKs (values()) of the `JdkVersion` enum now. The `AzureCloudFunctionSpec` verifies that unsupported JDK versions raise an exception. The problem is that it now supports all of them in the enum, and the test doesn't execute since it's data driven and no longer has any data (i.e. unsupported JDK).

My first thought was to make it `@PendingFeature` with a reason that it needs a unsupported JDK in the enum. But that doesn't work because PendingFeature says, "if every iteration of a data-driven test passes it will be reported as error". I can't use `groovy.test.NotYetImplemented` or the deprecated `groovy.transform.NotYetImplemented` for similar reasons. I also don't want to add to `JdkVersion` – e.g. JDK_19 – just to make a test work. That should be under it's unique PR if we want to do that anyway, IMO.

See: https://github.com/micronaut-projects/micronaut-starter/blob/f51cf6ff43ea7eca0aacbd8e7b2ff5ab8059f8be/starter-core/src/test/groovy/io/micronaut/starter/feature/function/azure/AzureCloudFunctionSpec.groovy#L27

File: starter-core/src/main/java/io/micronaut/starter/feature/function/azure/AzureFeatureValidator.java [hunk 1]
```diff
@@ -39,8 +38,11 @@ public void validatePostProcessing(Options options, ApplicationType applicationT
             if (options.getBuildTool() == BuildTool.GRADLE_KOTLIN) {
                 throw new IllegalArgumentException("The Azure Gradle plugin currently does not support the Kotlin Gradle DSL.");
             }
-            if (!(options.getJavaVersion() == JdkVersion.JDK_8 || options.getJavaVersion() == JdkVersion.JDK_11)) {
-                throw new IllegalArgumentException("The Azure Functions runtime only supports Java 8 or Java 11.");
+            switch (options.getJavaVersion()) {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/function/azure/AzureFeatureValidator.java]: remove

File: starter-core/src/test/groovy/io/micronaut/starter/feature/function/azure/AzureCloudFunctionSpec.groovy [hunk 2]
```diff
@@ -24,6 +33,8 @@ class AzureCloudFunctionSpec extends ApplicationContextSpec implements CommandOu
         where:
         [applicationType, jdkVersion, feature] << [
                 [ApplicationType.FUNCTION, ApplicationType.DEFAULT],
+                // TODO: this is an empty list until an unsupported JDK is added to JdkVersion,
+                //  and prevents test from executing with 'Data provider has no data' exception
                 ((JdkVersion.values() as List<JdkVersion>) - AzureFeatureValidatorSpec.supportedVersionsByAzureFunction()) as List<JdkVersion>,
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/function/azure/AzureCloudFunctionSpec.groovy]: replace with `JdkVersion.values()` and test for `noExceptionThrown()`

File: starter-core/src/test/groovy/io/micronaut/starter/feature/function/azure/AzureFeatureValidatorSpec.groovy [hunk 3]
```diff
@@ -33,6 +33,6 @@ class AzureFeatureValidatorSpec extends Specification {
     }
 
     static List<JdkVersion> supportedVersionsByAzureFunction() {
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/function/azure/AzureFeatureValidatorSpec.groovy]: delete

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: starter-core/src/test/groovy/io/micronaut/starter/feature/function/azure/AzureCloudFunctionSpec.groovy [hunk 4]
```diff
@@ -1,15 +1,24 @@
...
+import io.micronaut.starter.options.JdkVersion
+import io.micronaut.starter.options.Language
+import io.micronaut.starter.options.Options
+import io.micronaut.starter.options.TestFramework
+import spock.lang.PendingFeature
 import spock.lang.Unroll
 
 class AzureCloudFunctionSpec extends ApplicationContextSpec implements CommandOutputFixture {
 
+    // PendingFeature doesn't work here, because "if every iteration of a data-driven test passes it will be reported as error"
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/function/azure/AzureCloudFunctionSpec.groovy]: delete comment and `@PendingFeature`

**Review by sdelamo — CHANGES_REQUESTED** [review]

**Comment by sdelamo** [comment]
@wetted please, remember to place the PRs you want reviewed in in ready for review.

**Comment by wetted** [comment]
> @wetted please, remember to place the PRs you want reviewed in in ready for review.

Sorry, that was an oversight. I usually do that.

**Pull request merged**